1.3.0 on C* 1.2.19: cannot load cluster state (`nodes_status` remains null)

Hi,

I recently upgraded reaper from 1.2.2 to 1.3.0 and since then the UI is not able to load the cluster state anymore, it keeps spinning (it doesn't show neither the disk usage neither the nodes):

![spinning](https://user-images.githubusercontent.com/361403/47523062-70448480-d897-11e8-9d26-6f34212562a0.png)

Looking at the ajax calls made I can see that the endpoint `cluster/staging-cluster` returns a json object with `nodes_status` set to `null`:

![image](https://user-images.githubusercontent.com/361403/47523298-02e52380-d898-11e8-9ddc-72b507b69487.png)

In `cassandra-reaper.log` I can see that the following exception happens frequently:

```
Caused by: javax.management.InstanceNotFoundException: org.apache.cassandra.net:type=StreamManager
```

Note that I haven't checked the code (yet) so I don't know if this may be related or not.

Does it means that reaper 1.3.0 is not compatible with Cassandra 1.2 anymore?

Cheers,
Reynald

--- a/src/server/src/main/java/io/cassandrareaper/jmx/JmxProxyImpl.java
+++ b/src/server/src/main/java/io/cassandrareaper/jmx/JmxProxyImpl.java
@@ -113,7 +113,7 @@ final class JmxProxyImpl implements JmxProxy {
   private final ConcurrentMap<Integer, ExecutorService> repairStatusExecutors = Maps.newConcurrentMap();
   private final ConcurrentMap<Integer, RepairStatusHandler> repairStatusHandlers = Maps.newConcurrentMap();
   private final MetricRegistry metricRegistry;
-  private final StreamManagerMBean smProxy;
+  private final Optional<StreamManagerMBean> smProxy;
 
   private JmxProxyImpl(
       String host,
@@ -125,7 +125,7 @@ final class JmxProxyImpl implements JmxProxy {
       EndpointSnitchInfoMBean endpointSnitchMbean,
       FailureDetectorMBean fdProxy,
       MetricRegistry metricRegistry,
-      StreamManagerMBean smProxy) {
+      Optional<StreamManagerMBean> smProxy) {
 
     this.host = host;
     this.hostBeforeTranslation = hostBeforeTranslation;
@@ -174,7 +174,7 @@ final class JmxProxyImpl implements JmxProxy {
    *
    * @param handler Implementation of {@link RepairStatusHandler} to process incoming notifications
    *     of repair events.
-   * @param host hostname or ip address of Cassandra node
+   * @param originalHost hostname or ip address of Cassandra node
    * @param port port number to use for JMX connection
    * @param username username to use for JMX authentication
    * @param password password to use for JMX authentication
@@ -206,7 +206,7 @@ final class JmxProxyImpl implements JmxProxy {
       throw new ReaperException("Failure during preparations for JMX connection", e);
     }
     try {
-      Map<String, Object> env = new HashMap<>();
+      final Map<String, Object> env = new HashMap<>();
       if (username != null && password != null) {
         String[] creds = {username, password};
         env.put(JMXConnector.CREDENTIALS, creds);
@@ -218,9 +218,17 @@ final class JmxProxyImpl implements JmxProxy {
       StorageServiceMBean ssProxy
           = JMX.newMBeanProxy(mbeanServerConn, ObjectNames.STORAGE_SERVICE, StorageServiceMBean.class);
 
-      String cassandraVersion = ssProxy.getReleaseVersion();
+      final String cassandraVersion = ssProxy.getReleaseVersion();
       if (cassandraVersion.startsWith("2.0") || cassandraVersion.startsWith("1.")) {
         ssProxy = JMX.newMBeanProxy(mbeanServerConn, ObjectNames.STORAGE_SERVICE, StorageServiceMBean20.class);
+      }
+
+      final Optional<StreamManagerMBean> smProxy;
+      // StreamManagerMbean is only available since Cassandra 2.0
+      if (cassandraVersion.startsWith("1.")) {
+        smProxy = Optional.empty();
+      } else {
+        smProxy = Optional.of(JMX.newMBeanProxy(mbeanServerConn, ObjectNames.STREAM_MANAGER, StreamManagerMBean.class));
       }
 
       JmxProxy proxy
@@ -234,11 +242,13 @@ final class JmxProxyImpl implements JmxProxy {
               JMX.newMBeanProxy(mbeanServerConn, ObjectNames.ENDPOINT_SNITCH_INFO, EndpointSnitchInfoMBean.class),
               JMX.newMBeanProxy(mbeanServerConn, ObjectNames.FAILURE_DETECTOR, FailureDetectorMBean.class),
               metricRegistry,
-              JMX.newMBeanProxy(mbeanServerConn, ObjectNames.STREAM_MANAGER, StreamManagerMBean.class));
+              smProxy);
 
       // registering listeners throws bunch of exceptions, so do it here rather than in the constructor
       mbeanServerConn.addNotificationListener(ObjectNames.STORAGE_SERVICE, proxy, null, null);
-      mbeanServerConn.addNotificationListener(ObjectNames.STREAM_MANAGER, proxy, null, null);
+      if (smProxy.isPresent()) {
+        mbeanServerConn.addNotificationListener(ObjectNames.STREAM_MANAGER, proxy, null, null);
+      }
       LOG.debug("JMX connection to {} properly connected: {}", host, jmxUrl.toString());
 
       return proxy;
@@ -951,7 +961,7 @@ final class JmxProxyImpl implements JmxProxy {
     return cmProxy;
   }
 
-  StreamManagerMBean getStreamManagerMBean() {
+  Optional<StreamManagerMBean> getStreamManagerMBean() {
     return smProxy;
   }
 

--- a/src/server/src/main/java/io/cassandrareaper/jmx/JmxProxyImpl.java
+++ b/src/server/src/main/java/io/cassandrareaper/jmx/JmxProxyImpl.java
@@ -160,7 +160,7 @@ final class JmxProxyImpl implements JmxProxy {
     final HostAndPort hostAndPort = HostAndPort.fromString(host);
 
     return connect(
-          hostAndPort.getHostText(),
+          hostAndPort.getHost(),
           hostAndPort.getPortOrDefault(JMX_PORT),
           username,
           password,

--- a/src/server/src/main/java/io/cassandrareaper/jmx/StreamsProxy.java
+++ b/src/server/src/main/java/io/cassandrareaper/jmx/StreamsProxy.java
@@ -23,7 +23,7 @@ import java.util.Set;
 import javax.management.openmbean.CompositeData;
 
 import com.google.common.base.Preconditions;
-
+import com.google.common.collect.ImmutableSet;
 
 
 public final class StreamsProxy {
@@ -40,6 +40,10 @@ public final class StreamsProxy {
   }
 
   public Set<CompositeData> listStreams() {
-    return proxy.getStreamManagerMBean().getCurrentStreams();
+    if (proxy.getStreamManagerMBean().isPresent()) {
+      return proxy.getStreamManagerMBean().get().getCurrentStreams();
+    } else {
+      return ImmutableSet.of();
+    }
   }
 }

--- a/src/server/src/test/java/io/cassandrareaper/jmx/JmxProxyTest.java
+++ b/src/server/src/test/java/io/cassandrareaper/jmx/JmxProxyTest.java
@@ -19,6 +19,8 @@ package io.cassandrareaper.jmx;
 
 import io.cassandrareaper.ReaperException;
 
+import java.util.Optional;
+
 import javax.management.MBeanServerConnection;
 
 import com.google.common.base.Preconditions;
@@ -48,7 +50,7 @@ public final class JmxProxyTest {
 
   public static void mockGetStreamManagerMBean(JmxProxy proxy, StreamManagerMBean streamingManagerMBean) {
     Preconditions.checkArgument(proxy instanceof JmxProxyImpl, "only JmxProxyImpl is supported");
-    Mockito.when(((JmxProxyImpl)proxy).getStreamManagerMBean()).thenReturn(streamingManagerMBean);
+    Mockito.when(((JmxProxyImpl)proxy).getStreamManagerMBean()).thenReturn(Optional.of(streamingManagerMBean));
   }
 
   public static void mockGetEndpointSnitchInfoMBean(JmxProxy proxy, EndpointSnitchInfoMBean endpointSnitchInfoMBean) {

